src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    public static final String MESSAGE_WRONG_INDEX = "Wrong index";
+    public static final String MESSAGE_INPUT_IS_NULL = "The list is null";
+    private transient int size = 0;
+    private transient Node<T> first;
+    private transient Node<T> last;
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private Node<T> next;
+        private T element;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.element = element;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node;
+        if (isEmpty()) {
+            node = new Node<>(null, value, null);
+            first = node;
+        } else {
+            node = new Node<>(last, value, null);
+            last.next = node;
+        }
+        last = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkAddIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            final Node<T> nextNode = getNodeBy(index);
+            final Node<T> prevNode = nextNode.prev;
+            final Node<T> newNode = new Node<>(prevNode, value, nextNode);
+            if (prevNode != null) {
+                prevNode.next = newNode;
+            } else {
+                first = newNode;
+            }
+            nextNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T element : list) {
+                add(element);
+            }
+        } else {
+            throw new NullPointerException(MESSAGE_INPUT_IS_NULL);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkSearchIndex(index);
+        return getNodeBy(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkSearchIndex(index);
+        final Node<T> oldNode = getNodeBy(index);
+        T oldValue = oldNode.element;
+        oldNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkSearchIndex(index);
+        final Node<T> node = getNodeBy(index);
+        final Node<T> prevNode = node.prev;
+        final Node<T> nextNode = node.next;
+        final T value = node.element;
```

Comment:
try to avoid redundant variables where it's possible and assign a value when it's needed, same for all places

Reply:
Also, create a separate unlink() method 
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html